File: remove_duplicates_from_sorted_list.py
class ListNode:
    def __init__(self,x):
        self.val=x
        self.next=None
class Solution:
    def deleteDuplicates(self,head):     
        if head is None:
            return head
        pre=dummy=ListNode(1)
        pre.next=head
        current=head
        while(pre.next):
            while current.next and current.next.val==pre.next.val:
                current=current.next      
            else:
                if pre.next==current:
                    pre=pre.next
                    current=pre.next
                else:
                    pre.next=current.next
        return dummy.next
if __name__=='__main__':
    
    sol=Solution()
    
    head=l1=ListNode(1)
    lst1=[1,2,3,3,3,4,4,5,5,6,7,7]
    for element in lst1:
        l1.next=ListNode(element)
        l1=l1.next
        
    l1.next=None
    head1=head
    while head1:
        print head1.val,
        head1=head1.next
    list_head=sol.deleteDuplicates(head)
    print 
    while list_head:
        print list_head.val,
        list_head=list_head.next

File: Algorithm/CPP/remove_duplicates_from_sorted_list.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <stack>
# include <algorithm>
# include<time.h>
using namespace std;
struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* deleteDuplicates(ListNode* head) {
		auto dummy = new ListNode(-1);
		dummy->next = head;
		auto p = head;
		auto q = head->next;
		while (q)
		{
			if (p->val != q->val)
			{
				p = p->next;
				q = q->next;
			}
			else
			{
				p->next = q->next;
				auto tmp = q;
				delete tmp;
				q = p->next;
			}
		}
		return dummy->next;
	}
};
int main()
{
	ListNode *head1 = new ListNode(1);
	ListNode *head2 = new ListNode(-2);
	auto p = head1;
	auto q = head2;
	vector<int> vec{ 1,3,2,5,6};//2
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;

	Solution sol;

	auto pp = sol.deleteDuplicates(head1);
	while (pp != NULL)
	{
		cout << pp->val << " ";
		pp = pp->next;
	}
	cout << endl;
	system("pause");
	return 0;
}
